transcom/mymove
# B 19821 sit auth date to db int

Do not merge before https://github.com/transcom/mymove/pull/12593

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-19821)

## Summary

The point of this story is to store the SIT authorized end date value in the database so that it can be sent to Homesafe. Previously, this SIT authorized end date was just being calculated on the front end but not stored anywhere.

The sitStatus holds most of the frontend calculations, with `pkg/services/sit_status/shipment_sit_status.go` being where a lot of the meat and potatoes are for the SIT panel on the MTO page.

The work here captures that SIT authorized end date value (`sitStatus.CurrentSIT.SITAuthorizedEndDate`) and stores it in the `mto_shipments` table in either the `origin_sit_auth_end_date` column or the `dest_sit_auth_end_date` column, dependent on whether `sitStatus.CurrentSIT` is `Origin` or `Destination`. Logic was added to the handlers for updating a service item (when the TOO approves the service items, this populates the SIT panel initially), changing the SIT entry date of the service item, a SIT extension creation and approval by the TOO, and approval of a SIT extension request from the prime.

We also now return the SIT authorized end date stored on the shipment in the response for the prime v1, v2, and v3 `getMoveTaskOrder` endpoints. SIT entry date was already being included in the response for this endpoint at the service item level. Now, HSC will have both the SIT start date and SIT authorized end date available to them which is what the feature is all about.

### How to test

Origin:
1. As a customer, create a new move with shipments.
2. Counsel the move, approve as TOO
3. As the prime, request origin SIT. Make sure the SIT entry date is earlier than the current day you are testing AND the departure date is later than the current day.
4. On the MTO page, approve the origin SIT service items. If your SIT panel on the MTO page does not say "Current location: origin SIT" you can modify the SIT entry date on the service item below to be a date preceding the current day.
5. Check the `mto_shipments` table. You should see the `origin_sit_auth_end_date` value in the database populated with a value that matches what you see on the MTO page SIT panel for SIT authorized end date.
6. Edit the SIT authorized end date or the Total days of SIT approved fields and make sure if the SIT authorized end date changes, that change is also reflected in the database.
7. Call the prime `getMoveTaskOrder` endpoint for the move you created (prime v1, v2, or v3). Ensure that you see `originSitAuthEndDate` in the response, as well as an accurate `sitEntryDate` under the `DOFSIT` service item.

Request SIT Extension
8. As the prime, find the shipment you created for your new move that still has the current location as origin SIT.
9. "Request SIT Extension" as the prime
10. Go back to the MTO page as the TOO
11. Find the Review request next to "Additional Days Requested" 
![image](https://github.com/transcom/mymove/assets/147537467/4dadbf38-b06e-45ec-93b7-57cd7e07c4b2)
12. Reject the SIT extension. You can change the dates on the modal, but rejecting should NOT update the SIT auth end date or allowed days.
13. As the prime, request another SIT extension.
14. As the TOO, modify the SIT auth end date or days allowed and approve the extension.
15. Call the prime `getMoveTaskOrder` endpoint for the move you created (prime v1, v2, or v3). Ensure that you see `originSitAuthEndDate` in the response, as well as an accurate `sitEntryDate` under the `DOFSIT` service item. 

Destination:
1. As a customer, create a new move with shipments.
2. Counsel the move, approve as TOO
3. As the prime, request destination SIT. Make sure the SIT entry date is earlier than the current day you are testing AND the departure date is later than the current day.
4. On the MTO page, approve the destination SIT service items. If your SIT panel on the MTO page does not say "Current location: destination SIT" you can modify the SIT entry date on the service item below to be a date preceding the current day.
5. Check the `mto_shipments` table. You should see the `dest_sit_auth_end_date` value in the database populated with a value that matches what you see on the MTO page SIT panel for SIT authorized end date.
6. Edit the SIT authorized end date or the Total days of SIT approved fields and make sure if the SIT authorized end date changes, that change is also reflected in the database.
7. Call the prime `getMoveTaskOrder` endpoint for the move you created (prime v1, v2, or v3). Ensure that you see `destinationSitAuthEndDate` in the response, as well as an accurate `sitEntryDate` under the `DDFSIT ` service item.

Request SIT Extension
8. As the prime, find the shipment you created for your new move that still has the current location as destination SIT.
9. "Request SIT Extension" as the prime
10. Go back to the MTO page as the TOO
11. Find the Review request next to "Additional Days Requested" 
![image](https://github.com/transcom/mymove/assets/147537467/4dadbf38-b06e-45ec-93b7-57cd7e07c4b2)
12. Reject the SIT extension. You can change the dates on the modal, but rejecting should NOT update the SIT auth end date or allowed days.
13. As the prime, request another SIT extension.
14. As the TOO, modify the SIT auth end date or days allowed and approve the extension.
15. Call the prime `getMoveTaskOrder` endpoint for the move you created (prime v1, v2, or v3). Ensure that you see `originSitAuthEndDate` in the response, as well as an accurate `sitEntryDate` under the `DOFSIT` service item.
